tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -8882,6 +8883,84 @@ static NameLookup GetSingleFunctionDeclByName(clang::Sema *self, StringRef Name,
   return NameLookup{ pFoundDecl, nullptr };
 }
 
+namespace {
+/// A visitor class for validating HLSL attributes and annotations
+class HLSLAttributeChecker : public RecursiveASTVisitor<HLSLAttributeChecker> {
+public:
+  explicit HLSLAttributeChecker(Sema &S);
+
+  bool TraverseFunctionDecl(FunctionDecl* func);
+
+  bool VisitFunctionDecl(FunctionDecl *func);
+  bool VisitParmVarDecl(ParmVarDecl *parmVar);
+
+private:
+  typedef RecursiveASTVisitor<HLSLAttributeChecker> BaseType;
+
+  Sema &theSema;
+  clang::DiagnosticsEngine &theDiag;
+  const hlsl::ShaderModel *shaderModel;
+
+  bool isEntryFunction;
+};
+
+HLSLAttributeChecker::HLSLAttributeChecker(Sema &S)
+    : theSema(S), theDiag(S.getDiagnostics()), shaderModel(nullptr),
+      isEntryFunction(false) {
+  shaderModel =
+      hlsl::ShaderModel::GetByName(theSema.getLangOpts().HLSLProfile.c_str());
+}
+
+bool HLSLAttributeChecker::TraverseFunctionDecl(FunctionDecl* func) {
+  if (!VisitFunctionDecl(func))
+    return false;
+
+  for (auto* param : func->params())
+    if (!TraverseParmVarDecl(param))
+      return false;
+
+  // Avoid traversing the statements
+
+  return true;
+}
+
+bool HLSLAttributeChecker::VisitFunctionDecl(FunctionDecl *func) {
+  isEntryFunction = theSema.getLangOpts().HLSLEntryFunction == func->getName();
+
+  if (isEntryFunction && shaderModel->IsGS()) {
+    // Validate that GS has the maxvertexcount attribute
+    if (!func->hasAttr<HLSLMaxVertexCountAttr>()) {
+      theDiag.Report(func->getLocation(),
+                     diag::err_hlsl_missing_maxvertexcount_attr)
+          << func->getName();
+    }
+  }
+
+  return true;
+}
+
+bool HLSLAttributeChecker::VisitParmVarDecl(ParmVarDecl *parmVar) {
+  // Validate that stream-ouput objects are marked as inout
+  if (shaderModel->IsGS()) {
+    if (hlsl::IsHLSLStreamOutputType(parmVar->getType()) &&
+        !parmVar->hasAttr<HLSLInOutAttr>()) {
```

Comment:
Perhaps this validation could be applied without regard to the shader model.  Otherwise, it's being applied to all functions, but only if the current entry is a GS.

Reply:
Checking additional shader model specific requirements might be more difficult, since we will need to know whether this function is considered an entry, and which kind (esp. for lib target).

Reply:
I think this particular check could be globally applied in Sema::DiagnoseHLSLDecl.  Just need to move some HLSL type retrieval code out of if (hasSignSpec), then just before the end of the method, verify.
Here's a start:
```
diff --git a/tools/clang/lib/Sema/SemaHLSL.cpp b/tools/clang/lib/Sema/SemaHLSL.cpp
index 53d1f23d..9801de8e 100644
--- a/tools/clang/lib/Sema/SemaHLSL.cpp
+++ b/tools/clang/lib/Sema/SemaHLSL.cpp
@@ -10551,6 +10551,10 @@ bool Sema::DiagnoseHLSLDecl(Declarator &D, DeclContext *DC,
     }
   }

+  HLSLExternalSource *hlslSource = HLSLExternalSource::FromSema(this);
+  ArTypeObjectKind objKind = hlslSource->GetTypeObjectKind(qt);
+  ArBasicKind basicKind = hlslSource->GetTypeElementKind(qt);
+
   // Check for deprecated effect object type here, warn, and invalidate decl
   bool bDeprecatedEffectObject = false;
   bool bIsObject = false;
@@ -10609,9 +10613,6 @@ bool Sema::DiagnoseHLSLDecl(Declarator &D, DeclContext *DC,
   }

   if (hasSignSpec) {
-     HLSLExternalSource *hlslSource = HLSLExternalSource::FromSema(this);
-     ArTypeObjectKind objKind = hlslSource->GetTypeObjectKind(qt);
-     ArBasicKind basicKind = hlslSource->GetTypeElementKind(qt);
      // vectors or matrices can only have unsigned integer types.
      if (objKind == AR_TOBJ_MATRIX || objKind == AR_TOBJ_VECTOR || objKind == AR_TOBJ_BASIC || objKind == AR_TOBJ_ARRAY) {
          if (!IS_BASIC_UNSIGNABLE(basicKind)) {
@@ -10889,6 +10890,15 @@ bool Sema::DiagnoseHLSLDecl(Declarator &D, DeclContext *DC,
     }
   }

+  if (isParameter
+      && ( basicKind == ArBasicKind::AR_OBJECT_LINESTREAM ||
+           basicKind == ArBasicKind::AR_OBJECT_POINTSTREAM ||
+           basicKind == ArBasicKind::AR_OBJECT_TRIANGLESTREAM )
+      && !(usageIn && usageOut)) {
+    // TODO: diagnose
+    result = false;
+  }
+
   if (!result) {
     D.setInvalidType();
   }
```

Reply:
Good point! Thanks for pointing me to `DiagnoseHLSLDecl()`. Didn't notice it previously. Definitely much suitable to check there.